File: chapters/fundamental_algorithms/tree_traversal/code/c++/tree.cpp
struct node {
  std::vector<node> children;
  int id;
};

void dfs_recursive(const node& n) {
  // Here we are doing something...
  std::cout << n.id << '\n';

  for (auto const& child : n.children) {
    dfs_recursive(child);
  }
}

void DFS_recursive_postorder(const node& n) {
  for (auto const& child : n.children) {
    dfs_recursive_postorder(child);
  }

  // Here we are doing something...
  std::cout << n.id << '\n';
}

// This assumes only 2 children
void dfs_recursive_inorder_bnode(const node& n) {
  if (not n.children.empty()) {
    if (n.children.size() != 2) {
      std::cerr << "Not binary tree!\n";
      std::abort();
    }
    dfs_recursive_inorder_bnode(n.children[0]);
    std::cout << n.id << '\n';
    dfs_recursive_inorder_bnode(n.children[1]);
  } else {
    std::cout << n.id << '\n';
  }
}

void dfs_stack(const node& n) {
  std::stack<node> s;
  s.push(n);

  while (s.size() > 0) {
    node temp = s.top();
    s.pop();
    std::cout << s.top().id << '\n';

    for (auto const& child : temp.children) {
      s.push(child);
    }
  }
}

void bfs_queue(const node& n) {
  std::queue<node> q;
  q.push(n);

  while (q.size() > 0) {
    node temp = q.front();
    q.pop();

    std::cout << q.front().id << '\n';
    for (auto const& child : temp.children) {
      q.push(child);
    }
  }
}

File: chapters/fundamental_algorithms/tree_traversal/code/rust/tree.rs
use std::collections::VecDeque;

#[derive(Debug)]
struct Node {
    children: Vec<Node>,
    value: u64,
}

fn dfs_recursive(n: &Node) {
    println!("{}", n.value);

    for child in &n.children {
        dfs_recursive(child);
    }
}

fn dfs_stack(n: &Node) {
    let mut stack = vec![n];

    while let Some(current) = stack.pop() {
        println!("{}", current.value);
        stack.extend(&current.children);
    }
}

fn bfs_queue(n: &Node){
    let mut queue = VecDeque::new();
    queue.push_back(n);

    while let Some(current) = queue.pop_front() {
        println!("{}", current.value);
        queue.extend(&current.children);
    }
}

fn create_tree(num_row: u64, num_child: u64) -> Node {
    if num_row == 0 {
        return Node { children: vec![], value: 0 };
    }
    
    let children = (0..num_child)
        .map(|_| create_tree(num_row - 1, num_child))
        .collect();

    Node { children, value: num_row }
}

fn main() {
    let root = create_tree(3,3);
    println!("Recursive DFS:");
    dfs_recursive(&root);
    println!("Stack DFS:");
    dfs_stack(&root);
    println!("Queue BFS:");
    bfs_queue(&root);
}
